README.md
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.3339453.svg)](https://doi.org/10.5281/zenodo.3339453)
[![Docker Repository on quay.io](https://quay.io/repository/eschen42/galaxy-tardis/status "Docker Repository on quay.io")](https://quay.io/repository/eschen42/galaxy-tardis)

# TARDIS - Temporal Archive Remote Distribution and Installation System

## Motivation: Administering a Local Galaxy with Minimal Stress

Suppose that:
- You want to host a local instance of Galaxy.
- You want to lose nothing if your Galaxy is swallowed by a black hole.
- You want your backups to be efficient, easy to manage, secure, and offsite.

Basically, you want to be able to travel (back) in time.

The Galaxy ["Temporal Archive Remote Distribution and Installation System", https://github.com/HegemanLab/galaxy-tardis](https://github.com/HegemanLab/galaxy-tardis) may be right for you.
- Any resemblance of the Galaxy TARDIS to [the TARDIS from *Doctor Who*](https://en.wikipedia.org/wiki/TARDIS) is purely (albeit intentionally) coincidental.

The intent of Galaxy TARDIS is:
- to facilitate restoration of an existing ([`docker-galaxy-stable`](https://github.com/bgruening/docker-galaxy-stable/)-based) Galaxy instance to a known state.
- to facilitates migrating an instance to another host.

Notably, the intent is **not** to replace other automation systems (e.g., ansible); conversely, Galaxy TARDIS facilitate the job of automating backup and restoration of a `docker-galaxy-stable`-based Galaxy instance.

## Overview

- The purpose of the Galaxy `tardis` Docker image is to back up and restore Galaxy instances that are based on [docker-galaxy-stable](https://github.com/bgruening/docker-galaxy-stable/).
- The only storage back-end for backup implemented thus far is S3-compatible storage such as Ceph.
    - Because the implementation is modular, it should not be particularly challenging to support other back-ends.
- You can [pull the image](#how-to-pull-the-docker-image-by-tag) or [build it yourself](#how-to-build-the-docker-image).
- A practical example of application of the Galaxy TARDIS is given in the `restore example` directory.

### Documentation

Documentation and instructions for using the Galaxy TARDIS may be found at [https://hegemanlab.github.io/galaxy-tardis/tardis-intro.html](https://hegemanlab.github.io/galaxy-tardis/tardis-intro.html).
This also describes the illustrative example available in the `restore_example` directory.

### How to pull the Docker image by tag:

Identify a tag with a "Passed" security tag from [https://quay.io/repository/eschen42/galaxy-tardis?tab=tags](https://quay.io/repository/eschen42/galaxy-tardis?tab=tags).
> For example, when this was written, the "v0.0.1" had passed the "security scan" for known vulnerabilities.
  As time goes on, new vulnerabilites may be discovered and the "security scan" status of a tag may change.

Use the tag that you chose to pull the image and tag it as `tardis:latest`.  The `pull_tag.sh` exists to simplify this:
```bash
bash pull_tag.sh -s quay.io/eschen42/galaxy-tardis:v0.0.1 -t tardis:latest
```

### How to build the Docker image:

You can build the image with:
```bash
git clone https://github.com/HegemanLab/galaxy-tardis.git
cd galaxy-tardis
bash build_notar.sh
```

### Quick Start

Suppose that you have significant familiarity with the following (links are not provided to avoid the impression that following a few links is a substitute for significant experience):
- bash
- docker
- docker-compose
- S3
- s3cmd
- docker-galaxy-stable

If so, then try looking at the first lines of the comment at the head of the [`restore_example/DEMO` script](https://github.com/HegemanLab/galaxy-tardis/blob/master/restore_example/DEMO).
Otherwise, please [see Documentation](#documentation).

## Usage 

Please [see Documentation](#documentation) to understand how the parts of Galaxy TARDIS work.

If you were to:
- create `tags-for-tardis_envar-to-source.sh` by copying `tags-for-tardis_envar-to-source.sh.example`
- adjust it for your Galaxy instance (based on `docker-galaxy-stable` or one of its derivitives)
- and run:
```bash
source tardis_envar.sh
$TARDIS help
```
you would see the following help text:

```
tardis - Temporal Archive Remote Distribution and Installation System for Galaxy-in-Docker

Usage:
  tardis backup                - Back up PostgreSQL database and galaxy-central/config.
  tardis transmit              - Transmit datasets and backup to Amazon-S3-compatible storage.
  tardis cron [hour24UTC]      - Run backup and transmit daily at hour24 UTC.
  tardis restore_datasets      - Retrieve datasets from S3 (not desirable when using object store).
  tardis retrieve_config       - Retrieve database and config backup (but not datasets) from S3.
  tardis apply_config [date]   - Restore config from backup, whether from S3 or "tardis backup".
  tardis seed_database [date]  - Replace PostgreSQL database with copy from backup.
  tardis purge_empty_tmp_dirs  - Purge empty tmp directories that accumulate with datasets.
  tardis upgrade_database      - Upgrade the PostgreSQL database to match the Galaxy version.
  tardis bash                  - Enter a bash shell.
  tardis upgrade_conda {url_or_path} {md5sum}
                               - Upgrade conda (both arguments required)
where:
  date        - can be relative (e.g., "1 hour ago") or absolute (e.g., any format accepted by the
                  --date argument of the Linux `date` program, see e.g. https://linux.die.net/man/1/date)
  hour24UTC   - any two digit hour for backup to occur; must be UTC (GMT), not local time.
  url_or_path - any URL from https://repo.continuum.io/miniconda/, or path (e.g., if you
                  copied the miniconda installer to your export directory)
  md5sum      - MD5 digest for url_or_path, e.g., from https://repo.continuum.io/miniconda/

Optional environment variable:
  PGDATA_SUBDIR (default "main") - name of subdirectory of PGDATA_PARENT where PostgreSQL database lives
    - Used by "seed database"
Required environment Variables (set using the "-e" option of the "docker run")
  These are the environment variables and the tasks that require them:
    EXPORT_DIR (usually "/export") - path to directory containing "galaxy-central"
      - Used by most tasks
    HOST_EXPORT_DIR - host path to the EXPORT_DIR as bind-mounted in docker
      - Used by "seed database"
    HOST_PGDATA_PARENT - host directory whose PGDATA_SUBDIR subdirectory is
        bind-mounted by the "galaxy-postgres" container to the path specified by PGDATA
      - Used by "seed database"
    PGDATA_PARENT - bind-mount within "tardis" for HOST_PGDATA_PARENT
      - Used by "seed database"
    IMAGE_POSTGRES - docker image for PostgreSQL, e.g., "quay.io/bgruening/galaxy-postgres"
      - Used by "seed database"
    TAG_POSTGRES - tag for docker image for PostgreSQL, e.g., "9.6.5_for_19.01"
      - Used by "seed database"
    IMAGE_GALAXY_INIT - docker image for initializing Galaxy, e.g., "quay.io/bgruening/galaxy-init"
      - Used by "apply_config" and "upgrade_conda"
    TAG_GALAXY - tag for docker image for PostgreSQL, e.g., "19.01"
      - Used by "apply_config" and "upgrade_conda"
    CONTAINER_GALAXY - name of instantiated PostgreSQL container, e.g., "galaxy-init"
      - Used by "apply_config" and "upgrade_conda"
Required bind-mounts:
  "/export"       - required by all but "bash" and "help"
  "/var/run/docker.sock"  - required by "seed_database", "backup"
  "/opt/s3/dest.s3cfg"    - required by "transmit", "retrieve_config", and "restore_datasets"
  "/opt/s3/dest.config"   - required by "transmit", "retrieve_config", and "restore_datasets"

```

## Supporting Software

### CVS

Configuration is backed up using CVS repositories

#### Why use CVS rather than Git?

CVS (Concurrent Versions System, [https://www.nongnu.org/cvs/](https://www.nongnu.org/cvs/)) stores all revisions of a text file in an extremely compact format.  This project backs up the Galaxy database to a single SQL file.  Multiple revisions of this file take up an much larger space in a Git repository, whereas, in a CVS repository, they take up little more room than a few times the size of the SQL file.  CVS has been replaced for general software development, but it seems to fill a good niche here.  On the other hand, a more compelling question might be "Why use CVS rather than RCS?", since both CVS and RCS use the same storage format - maybe CVS is more familiar and easier to install (CVS has a single binary, compared to nine for RCS).

There may be a modern source control system that could achieve compact storage and a single binary.  Fossil ([https://www.fossil-scm.org/](https://www.fossil-scm.org/)) seems like a possible candidate, but I have not yet worked with it enough to become familiar with its operation and security model.

#### Statically linked `cvs` binary

The `support/cvs-static` binary, version 1.12.13, was compiled and statically linked as described at [https://github.com/eschen42/cvs-static#how-to-build-a-statically-linked-cvs-binary](https://github.com/eschen42/cvs-static#how-to-build-a-statically-linked-cvs-binary)

### Busybox

This Docker image is based on Alpine linux, which uses the `musl` C library rather than `glibc`.  Busybox provides diverse functionality, especially useful inside a very minimal Docker container.  However, I don't want to have track one `musl` busybox and another `glibc` busybox, so I replaced the Alpine busybox in the image with a statically linked busybox, built as described at [https://github.com/eschen42/alpine-cbuilder#statically-linked-busybox](https://github.com/eschen42/alpine-cbuilder#statically-linked-busybox).

### Docker clients

The `support/docker-usernetes` binary is a statically linked binary that was extracted from:
[https://github.com/rootless-containers/usernetes/releases/tag/v20190511.1](https://github.com/rootless-containers/usernetes/releases/tag/v20190511.1)
as follows:
```bash
wget -O - https://github.com/rootless-containers/usernetes/releases/download/v20190511.1/SHA256SUM 2>/dev/null
wget https://github.com/rootless-containers/usernetes/releases/download/v20190511.1/usernetes-x86_64.tbz
sha256sum usernetes-x86_64.tbz
# If the sha256 sums don't match, stop here!
bzip2 -d usernetes-x86_64.tbz
tar -xvf usernetes-x86_64.tar usernetes/bin/docker
cp usernetes/bin/docker support/docker-usernetes
rm -rf usernetes-x86_64.tar usernetes
```

### dcron - Dillon's Cron

[Dillon's Cron Daemon](http://www.jimpryor.net/linux/dcron.html) was chosen over the more traditional Vixie Cron because:
- "having to combine a cron daemon with another daemon like anacron makes for too much complexity", and
- "All jobs are run with `/bin/sh` for conformity and portability".

## Bugs and missing features

See [./RELEASES.md#bugs-and-missing-features](./RELEASES.md#bugs-and-missing-features)
 
## Release notes

See [./RELEASES.md#release-notes](./RELEASES.md#release-notes)


RELEASES.md
## Bugs and missing features

- The approach of rebuilding conda environments is sensitive to changes of package availability in conda channels.
  - A more robust alternative would be to backup tarballs of the conda environments themselves.
- Need to back up and restore:
  - static HTML content ("welcome")
  - other features that I don't use yet
- Security issues
  - As of 2019.08.12, [CVE-2019-14697](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2019-14697) is still not fixed in the latest stable Alpine Linux build, 3.10.1.

## Release notes

- [v0.0.2](https://github.com/HegemanLab/galaxy-tardis/compare/v0.0.3...v0.0.2)
  - Upgrade Alpine Linux to 3.10.1
- [v0.0.2](https://github.com/HegemanLab/galaxy-tardis/compare/v0.0.2...v0.0.1)
  - Upgrade static CVS binary to 1.12.13
  - Upgrade static busybox binary to 1.31.0
- v0.0.1
  - First release


pull_tag.sh
# Created by argbash-init v2.5.1
#!/bin/bash

# ARG_OPTIONAL_SINGLE([srcimgtag],[s],[source path and tag to tardis image],[missing])
# ARG_OPTIONAL_SINGLE([tag],[t],[final tag],[tardis:latest])
# ARGBASH_SET_INDENT([  ])
# ARG_HELP([Pull a docker image and tag it (default tag=tardis:latest)],[e.g.:\nbash pull_tag.sh -s quay.io/eschen42/galaxy-tardis:v0.0.3\nbash pull_tag.sh -s quay.io/eschen42/galaxy-tardis:v0.0.3 -t tardis:v0.0.3])
# ARGBASH_GO()
# needed because of Argbash --> m4_ignore([
### START OF CODE GENERATED BY Argbash v2.8.1 one line above ###
# Argbash is a bash code generator used to get arguments parsing right.
# Argbash is FREE SOFTWARE, see https://argbash.io for more info
# Generated online by https://argbash.io/generate


# # When called, the process ends.
# Args:
#   $1: The exit message (print to stderr)
#   $2: The exit code (default is 1)
# if env var _PRINT_HELP is set to 'yes', the usage is print to stderr (prior to $1)
# Example:
#   test -f "$_arg_infile" || _PRINT_HELP=yes die "Can't continue, have to supply file as an argument, got '$_arg_infile'" 4
die()
{
  local _ret=$2
  test -n "$_ret" || _ret=1
  test "$_PRINT_HELP" = yes && print_help >&2
  echo "$1" >&2
  exit ${_ret}
}


# Function that evaluates whether a value passed to it begins by a character
# that is a short option of an argument the script knows about.
# This is required in order to support getopts-like short options grouping.
begins_with_short_option()
{
  local first_option all_short_options='sth'
  first_option="${1:0:1}"
  test "$all_short_options" = "${all_short_options/$first_option/}" && return 1 || return 0
}

# THE DEFAULTS INITIALIZATION - OPTIONALS
_arg_srcimgtag="missing"
_arg_tag="tardis:latest"


# Function that prints general usage of the script.
# This is useful if users asks for it, or if there is an argument parsing error (unexpected / spurious arguments)
# and it makes sense to remind the user how the script is supposed to be called.
print_help()
{
  printf '%s\n' "Pull a docker image and tag it (default tag=tardis:latest)"
  printf 'Usage: %s [-s|--srcimgtag <arg>] [-t|--tag <arg>] [-h|--help]\n' "$0"
  printf '\t%s\n' "-s, --srcimgtag: source path and tag to tardis image (default: 'missing')"
  printf '\t%s\n' "-t, --tag: final tag (default: 'tardis:latest')"
  printf '\t%s\n' "-h, --help: Prints help"
  printf '\n%s\n' "e.g.:
bash pull_tag.sh -s quay.io/eschen42/galaxy-tardis:v0.0.3
bash pull_tag.sh -s quay.io/eschen42/galaxy-tardis:v0.0.3 -t tardis:v0.0.3"
}


# The parsing of the command-line
parse_commandline()
{
  while test $# -gt 0
  do
    _key="$1"
    case "$_key" in
      # We support whitespace as a delimiter between option argument and its value.
      # Therefore, we expect the --srcimgtag or -s value.
      # so we watch for --srcimgtag and -s.
      # Since we know that we got the long or short option,
      # we just reach out for the next argument to get the value.
      -s|--srcimgtag)
        test $# -lt 2 && die "Missing value for the optional argument '$_key'." 1
        _arg_srcimgtag="$2"
        shift
        ;;
      # We support the = as a delimiter between option argument and its value.
      # Therefore, we expect --srcimgtag=value, so we watch for --srcimgtag=*
      # For whatever we get, we strip '--srcimgtag=' using the ${var##--srcimgtag=} notation
      # to get the argument value
      --srcimgtag=*)
        _arg_srcimgtag="${_key##--srcimgtag=}"
        ;;
      # We support getopts-style short arguments grouping,
      # so as -s accepts value, we allow it to be appended to it, so we watch for -s*
      # and we strip the leading -s from the argument string using the ${var##-s} notation.
      -s*)
        _arg_srcimgtag="${_key##-s}"
        ;;
      # See the comment of option '--srcimgtag' to see what's going on here - principle is the same.
      -t|--tag)
        test $# -lt 2 && die "Missing value for the optional argument '$_key'." 1
        _arg_tag="$2"
        shift
        ;;
      # See the comment of option '--srcimgtag=' to see what's going on here - principle is the same.
      --tag=*)
        _arg_tag="${_key##--tag=}"
        ;;
      # See the comment of option '-s' to see what's going on here - principle is the same.
      -t*)
        _arg_tag="${_key##-t}"
        ;;
      # The help argurment doesn't accept a value,
      # we expect the --help or -h, so we watch for them.
      -h|--help)
        print_help
        exit 0
        ;;
      # We support getopts-style short arguments clustering,
      # so as -h doesn't accept value, other short options may be appended to it, so we watch for -h*.
      # After stripping the leading -h from the argument, we have to make sure
      # that the first character that follows coresponds to a short option.
      -h*)
        print_help
        exit 0
        ;;
      *)
        _PRINT_HELP=yes die "FATAL ERROR: Got an unexpected argument '$1'" 1
        ;;
    esac
    shift
  done
}

# Now call all the functions defined above that are needed to get the job done
parse_commandline "$@"

# OTHER STUFF GENERATED BY Argbash

### END OF CODE GENERATED BY Argbash (sortof) ### ])
# [ <-- needed because of Argbash
if [ $_arg_srcimgtag != missing ]; then
  docker pull $_arg_srcimgtag && docker tag $_arg_srcimgtag $_arg_tag
  exit $?
else
  print_help
  exit 1
fi
# vim: sw=2 ts=2 et ai :
# ] <-- needed because of Argbash


restore_example/compose_restore.sh
#!/bin/bash
usage() {
  echo "Usage (note that options may be combined):"
  echo "  $0 --help                   Show this help text"
  echo "  $0 --datasets               Restore ALL dataset file (not necessary when using Galaxy object store)"
  echo "  $0 --retrieve_config        Retrieve config data"
  echo "  $0 --apply_config ['date']  Apply retrieved config data"
  echo "  $0 --database ['date']      Restore PostgreSQL database (this makes database changes since last backup unavailable)"
  echo "        'date' can be used to specify the newest backup can be applied, i.e., exclude backups newer than the date"
  echo "        'date' can be any format accepted by the Linxx 'date' program (https://linux.die.net/man/1/date),"
  echo "            which may be relative (e.g., '3 days ago')"
  echo "            or RFC822  (e.g., '28 Apr 2019 05:20:23 -5')  [where '-5' means five hours behind UTC]"
  echo "            or pseudo ISO8601 ('2019-04-28 05:20:23 -05:00')  [where '-05:00' means five hours behind UTC]."
  echo "            It is quite flexible, so go ahead and try something that makes sense to you."
  echo "  $0 --db_upgrade             Upgrade PostgreSQL to match installed Galaxy version"
  echo "  $0 --Miniconda3             Upgrade /export/_conda to the latest Miniconda3 version. (experimental, requires lynx)"
  echo "  $0 --Miniconda2             Upgrade /export/_conda to the latest Miniconda2 version. (experimental, requires lynx)"
  echo "        Upgrading to Miniconda2 will pointless once Galaxy has sunseted Python 2."
}
if [ -z "$1" ]; then
  usage
  exit 1
fi

APPLY_CONFIG=false
RETRIEVE_CONFIG=false
RESTORE_DATASETS=false
RESTORE_DATABASE=false
UPGRADE_DATABASE=false
UPGRADE_MINICONDA3=false
UPGRADE_MINICONDA2=false
last_arg=""
last_postgres=""
for arg in "$@"; do
  echo "arg: ${arg}"
  if [ ${arg:0:2} == "--" ]; then
   case "${arg:2}" in
      Miniconda3)
        echo "  Upgrade /export/_conda to the latest Miniconda3 version. (experimental, requires lynx)"
        UPGRADE_MINICONDA3=true
        ;;
      Miniconda2)
        echo "  Upgrade /export/_conda to the latest Miniconda2 version. (experimental, requires lynx)"
        UPGRADE_MINICONDA2=true
        ;;
      db_upgrade)
        echo "  Restore PostgreSQL database (this makes database changes since last backup unavailable)"
        UPGRADE_DATABASE=true
        ;;
      database)
        echo "  Restore PostgreSQL database (this makes database changes since last backup unavailable)"
        RESTORE_DATABASE=true
        ;;
      datasets)
        echo "  Restore ALL dataset file (not necessary when using Galaxy object store)"
        RESTORE_DATASETS=true
        ;;
      retrieve_config)
        echo "  Retrieve config data"
        RETRIEVE_CONFIG=true
        ;;
      apply_config)
        echo "  Apply retrieved config data"
        APPLY_CONFIG=true
        ;;
      help)
        usage
        exit 0
        ;;
      *)
        echo "Unrecognized option ${arg}"
        usage
        exit 1
        ;;
    esac
  elif [ "${last_arg}" == "--database"  ]; then
    date -Iseconds --date="${arg}"
    my_date=$( date -Iseconds --date="${arg}" ) || {
      echo "Invalid date format: ${arg}"
      echo "You may want to experiment with running:"
      echo "    date -Iseconds --date='your date and time string here'"
      exit 1
    }
    last_postgres="${arg}"
    echo "    Database rollback date to $my_date"
  elif [ "${last_arg}" == "--apply_config"  ]; then
    date -Iseconds --date="${arg}"
    my_date=$( date -Iseconds --date="${arg}" ) || {
      echo "Invalid date format: ${arg}"
      echo "You may want to experiment with running:"
      echo "    date -Iseconds --date='your date and time string here'"
      exit 1
    }
    last_config="${arg}"
    echo "    Configuration rollback date to $my_date"
  else
    echo "Unrecognized option '${arg}'"
    usage
    exit 1
  fi
  last_arg="${arg}"
done

echo "
APPLY_CONFIG=$APPLY_CONFIG
RETRIEVE_CONFIG=$RETRIEVE_CONFIG
RESTORE_DATASETS=$RESTORE_DATASETS
RESTORE_DATABASE=$RESTORE_DATABASE
UPGRADE_DATABASE=$UPGRADE_DATABASE
UPGRADE_MINICONDA3=$UPGRADE_MINICONDA3
UPGRADE_MINICONDA2=$UPGRADE_MINICONDA2
"
# Set the actual script directory per https://stackoverflow.com/a/246128
SOURCE="${BASH_SOURCE[0]}"
while [ -h "$SOURCE" ]; do # resolve $SOURCE until the file is no longer a symlink
  DIR="$( cd -P "$( dirname "$SOURCE" )" >/dev/null 2>&1 && pwd )"
  SOURCE="$(readlink "$SOURCE")"
  [[ $SOURCE != /* ]] && SOURCE="$DIR/$SOURCE" # if $SOURCE was a relative symlink, we need to resolve it relative to the path where the symlink file was located
done
DIR="$( cd -P "$( dirname "$SOURCE" )" >/dev/null 2>&1 && pwd )"
echo "DIR is $DIR"
pushd $DIR

#set -x

# Get invariant environment variables for this Galaxy instance
source env-for-compose-to-source.sh

# Get tags produced by build-orchestration-images.sh
source tags-for-compose-to-source.sh

SAVE_DIR=$DIR
pushd tardis
. tardis_envar.sh
popd # undo `pushd tardis`
DIR=$SAVE_DIR

if [ -z "$TARDIS" ]; then
  echo "FATAL - The TARDIS environment variable was not set; review tardis_envar.sh to see what variables you might need to set."
  exit 1
fi

if [ -z "$EXPORT_DIR" ]; then
  echo "FATAL - The EXPORT_DIR environment variable was not set; review tardis_envar.sh to see what variables you might need to set."
  exit 1
fi

echo EXPORT_DIR=$EXPORT_DIR
if [ ! -e $EXPORT_DIR/backup ]; then
  rootlesskit --disable-host-loopback mkdir -p $EXPORT_DIR/backup
fi

# Upgrade /export/_conda to latest Miniconda2 or Miniconda3
if [ "$UPGRADE_MINICONDA3" != "false" -o "$UPGRADE_MINICONDA2" != "false" ]; then
  which lynx || {
    echo '
      Upgrading to latest Miniconda2 or Miniconda3 requires that "lynx" be installed.
      Alternatively, you can visit https://repo.continuum.io/miniconda/ to find what
      you want and then run something like:
        $TARDIS upgrade_conda https://repo.continuum.io/miniconda/Miniconda2-4.6.14-Linux-x86_64.sh  faa7cb0b0c8986ac3cacdbbd00fe4168
    '
    exit 1
  }
  # The following is approximately equivalent to:
  #   lynx -nonumbers -width=160 -dump https://repo.continuum.io/miniconda/ | grep -v https |  grep latest-Linux-x86_64 | sed -e 's/^[ ]*\([^ ]*\).* \([^ ]*\)$/'"$TARDIS"' upgrade_conda \1 \2/'
  #   $TARDIS upgrade_conda https://repo.continuum.io/miniconda/Miniconda2-4.6.14-Linux-x86_64.sh  faa7cb0b0c8986ac3cacdbbd00fe4168
  my_substring=Miniconda2
  if [ "$UPGRADE_MINICONDA3" != "false" ]; then
    my_substring=Miniconda3
  fi
  my_command() {
    lynx -nonumbers -width=160 -dump https://repo.continuum.io/miniconda/ | \
      grep -v https | \
      grep $my_substring | \
      grep latest-Linux-x86_64 | \
      sed -e 's@^[ ]*\([^ ]*\).* \([^ ]*\)$@'"$TARDIS"' upgrade_conda https://repo.continuum.io/miniconda/\1 \2@'
  }
  # This is a very ugly way to achieve what I want to do, but at present I cannot get a functional alternative to work.
  #   This is equivalent to running something like
  #     $TARDIS upgrade_conda https://repo.continuum.io/miniconda/Miniconda2-4.6.14-Linux-x86_64.sh  faa7cb0b0c8986ac3cacdbbd00fe4168
  my_command > $DIR/my_docker_command
  . $DIR/my_docker_command
  rm $DIR/my_docker_command
fi

if [  "$RESTORE_DATABASE" != "false" -o "$APPLY_CONFIG" != "false" ]; then
  # ensure that none of the containers in the compose file are instantiated
  echo "DIR is $DIR"
  $DIR/compose_stop.sh
  docker-compose -f $COMPOSE_FILE ps | grep galaxy-postgres || rootlesskit --disable-host-loopback bash -c "
  if [ -f $PGDATA_DIR/postmaster.pid ]; then
    rm $PGDATA_DIR/postmaster.pid
  fi
"
fi

### Retrieve data for export/config if requested, but don't apply it yet ###

SUCCESS=YES
if [ "$RETRIEVE_CONFIG" != "false" ]; then
  ($TARDIS retrieve_config && echo configuration retrieval succeeded) || SUCCESS=NO
  if [ "$SUCCESS" == "NO" ]; then
    echo "Configuration retrieval did not succeed"
    exit 1
  fi
else
  echo skipping configuration retrieval
fi

### Apply previously-retrieved data for export/config ###

SUCCESS=YES
if [ "$APPLY_CONFIG" != "false" ]; then
  if [ ! -d $EXPORT_DIR/backup/config ]; then
    echo "WARNING missing backup - $EXPORT_DIR/backup/config not found"
    exit 1
  fi
  ($TARDIS apply_config "${last_config}" && echo configuration application succeeded) || SUCCESS=NO
  if [ "$SUCCESS" == "NO" ]; then
    echo "Configuration application did not succeed"
    exit 1
  fi
else
  echo skipping configuration application
fi

### Retrieve and apply dataset data ###

SUCCESS=YES
if [ "$RESTORE_DATASETS" != "false" ]; then
  ($TARDIS restore_datasets && echo dataset-file restoration succeeded) || SUCCESS=NO
  if [ "$SUCCESS" == "NO" ]; then
    echo Dataset-file restoration failed or skipped
    exit 1
  fi
else
  echo skipping dataset-file restoration
fi

### Apply previously-retrieved database data ###

if [ "$RESTORE_DATABASE" != "false" ]; then
  ($TARDIS seed_database "${last_postgres}" && echo dataset-file restoration succeeded) || SUCCESS=NO
  if [ "$SUCCESS" == "NO" ]; then
    echo Database restoration failed or skipped
    exit 1
  fi
fi

# sh manage_db.sh upgrade
if [ "$UPGRADE_DATABASE" != "false" ]; then
  ($TARDIS upgrade_database && echo dataset-file restoration succeeded) || SUCCESS=NO
  if [ "$SUCCESS" == "NO" ]; then
    echo Database upgrade failed or skipped
    exit 1
  fi
fi

popd  # undo `pushd $DIR`

exit 0


tardis_envar.sh
#!/bin/bash source this file - do not execute it

# set the actual script directory per https://stackoverflow.com/a/246128
SOURCE="${BASH_SOURCE[0]}"
while [ -h "$SOURCE" ]; do # resolve $SOURCE until the file is no longer a symlink
  DIR="$( cd -P "$( dirname "$SOURCE" )" >/dev/null 2>&1 && pwd )"
  SOURCE="$(readlink "$SOURCE")"
  [[ $SOURCE != /* ]] && SOURCE="$DIR/$SOURCE" # if $SOURCE was a relative symlink, we need to resolve it relative to the path where the symlink file was located
done
DIR="$( cd -P "$( dirname "$SOURCE" )" >/dev/null 2>&1 && pwd )"

source $DIR/tags-for-tardis_envar-to-source.sh
# tags-for-tardis_envar-to-source.sh may contain, e.g.:
# TAG_GALAXY='19.01'
# EXPORT_DIR=/home/demo/export
# INTERNAL_EXPORT_DIR='/export'
# PGDATA_PARENT=/home/demo/postgres
# IMAGE_POSTGRES='quay.io/bgruening/galaxy-postgres'
# TAG_POSTGRES=9.6.5_for_19.01
# IMAGE_GALAXY_INIT='quay.io/bgruening/galaxy-init'
# CONTAINER_GALAXY_INIT='galaxy-init'
# TAG_GALAXY='19.01'

if [ ! -d ${EXPORT_DIR:?} ]; then
  # fail if EXPORT_DIR is not specified; to address this failure, e.g., EXPORT_DIR=/full/path/to/export
  echo "Please set EXPORT_DIR (the directory which must contain the exported files referenced by galomix-compose) before sourcing ${SOURCE}"
elif [ ! -d ${PGDATA_PARENT:?} ]; then
  # fail if PGDATA_PARENT is not specified; to address this failure, e.g., PGDATA_PARENT=/mnt/ace/piscesv/postgres
  echo "Please set PGDATA_PARENT (the directory which must contain a directory named 'main' containing the PostgreSQL data for galomix-compose) before sourcing ${SOURCE}"
elif [ -z "${TAG_POSTGRES:?}" ]; then
  # fail if TAG_POSTGRES is not specified; to address this failure, e.g., TAG_POSTGRES="9.6.5_for_19.01"
  echo "Please set TAG_POSTGRES (a valid tag for an image of quay.io/bgruening/galaxy-postgres) before sourcing ${SOURCE}"
elif [ -z "${MY_S3CFG}" ]; then
  echo "ERROR: environment variable MY_S3CFG is not set"
elif [ ! -f ${MY_S3CFG:?} ]; then
  # fail if ${MY_S3CFG:?} is not properly set or is not a path to a file
  echo "ERROR: ${MY_S3CFG:?} does not exist or is not a file"
elif [ ! -f $DIR/s3/dest.config ]; then
  # fail if $DIR/s3/dest.config is not proper
  echo "ERROR: $DIR/s3/dest.config does not exist or is not a file"
else
  TARDIS="docker run --rm -ti \
    -v ${XDG_RUNTIME_DIR:?}/docker.sock:/var/run/docker.sock \
    -v ${MY_S3CFG:?}:/opt/s3/dest.s3cfg \
    -v $DIR/s3/dest.config:/opt/s3/dest.config \
    -v ${EXPORT_DIR:?}:${INTERNAL_EXPORT_DIR:?} \
    -e EXPORT_DIR=${INTERNAL_EXPORT_DIR:?} \
    -e HOST_EXPORT_DIR=${EXPORT_DIR:?} \
    -v ${PGDATA_PARENT:?}:/pgparent \
    -e PGDATA_PARENT=/pgparent \
    -e HOST_PGDATA_PARENT=${PGDATA_PARENT:?} \
    -e PGDATA_SUBDIR=${PGDATA_SUBDIR:-main} \
    -e TAG_POSTGRES=${TAG_POSTGRES:?} \
    -e IMAGE_POSTGRES=${IMAGE_POSTGRES:?} \
    -e TAG_GALAXY=${TAG_GALAXY:?} \
    -e IMAGE_GALAXY_INIT=${IMAGE_GALAXY_INIT:?} \
    -e CONTAINER_GALAXY_INIT=${CONTAINER_GALAXY_INIT:?} \
    --name tardis tardis"
  echo "TARDIS=$TARDIS"
fi
